rmw_zenoh_cpp/src/rmw_service.cpp

```diff
@@ -0,0 +1,419 @@
+  service_data->zn_request_subscriber_ = zn_declare_subscriber(
+    service_data->zn_session_,
+    service_data->zn_request_topic_key_,
+    zn_subinfo_default(),  // NOTE(CH3): Default for now
+    service_data->zn_request_sub_callback
+  );
+
+  // Init Zenoh queryable for availability checking
+  service_data->zn_queryable_ = zn_declare_queryable(
+    s, service->service_name, EVAL, [](ZNQuery * query){
+      const zn_string * resource = zn_query_res_name(query);
+      const zn_string * predicate = zn_query_predicate(query);
+
+      std::string key(resource->val, resource->len);
+      std::string response("available");  // NOTE(CH3): The contents actually don't matter...
+
+      zn_send_reply(query, key.c_str(), (const unsigned char *)response.c_str(), response.length());
+    }
+  );
+  if (service_data->zn_queryable_ == 0) {
+    RMW_SET_ERROR_MSG("failed to create availability queryable for service");
+
+    zn_undeclare_subscriber(service_data->zn_request_subscriber_);
+
+    allocator->deallocate(service_data->request_type_support_, allocator->state);
+    allocator->deallocate(service_data->response_type_support_, allocator->state);
+    allocator->deallocate(service->data, allocator->state);
+    allocator->deallocate(service, allocator->state);
+    return nullptr;
+  }
+
+  return service;
+}
+
+/// DESTROY SERVICE SERVER =====================================================
+// Destroy and deallocate an RMW service server
+rmw_ret_t
+rmw_destroy_service(rmw_node_t * node, rmw_service_t * service)
+{
+  RCUTILS_LOG_INFO_NAMED("rmw_zenoh_cpp", "[rmw_destroy_service] %s", service->service_name);
+
+  // ASSERTIONS ================================================================
+  RMW_CHECK_ARGUMENT_FOR_NULL(node, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(service, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    node,
+    node->implementation_identifier,
+    eclipse_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    service,
+    service->implementation_identifier,
+    eclipse_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  // OBTAIN ALLOCATOR ==========================================================
+  rcutils_allocator_t * allocator = &node->context->options.allocator;
+
+  // CLEANUP ===================================================================
+  auto service_data = static_cast<rmw_service_data_t *>(service->data);
+
+  zn_undeclare_subscriber(service_data->zn_request_subscriber_);
+  zn_undeclare_queryable(service_data->zn_queryable_);
+
+  allocator->deallocate(service_data->request_type_support_,
+                        allocator->state);
+  allocator->deallocate(service_data->response_type_support_,
+                        allocator->state);
+  allocator->deallocate(service->data, allocator->state);
+  allocator->deallocate(service, allocator->state);
```

Comment:
Does this leak `service->service_name`?

Reply:
Yes

Updated: 23497d4430f890dc62fb9995c66fb2be8602b51e

(Sorry for the massive commit, it's kind of intertwined...)

Reply:
Ja

Updated: 23497d4

(Sorry for the massive commit, it's kind of intertwined...)